Add Partial Flashing documentation (#65)
lancaster-university/codal-microbit-v2: docs/bluetooth/MicroBitPartialFlashing.md

# Partial Flashing Service

## Introduction

The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the firmware. For MakeCode programs this is the user's script, and for Python programs this is the file system.

This service exists in partnership with the DFU service that allows a client to fully flash the micro:bit.

Diagrams showing the flow of data during partial flashing can be found at [the bottom of the document](#partial flashing client).

## Bluetooth Service Specification

> TODO: Does BLE profile exist for v2?

Reviewer: For V2 we're confirming to the same profile spec as we did for V1, so linking to same place is fine. We haven't got an updated version.
Do we need to modify that spec to update it for the small variation between V1 and V2?

Author: The DFU service has changed to the up to date Nordic DFU

## Enabling the service

> Currently partial flashing is supported by programs built using MakeCode or the Python editor. This is because the service requires the user program to be appended onto the end of CODAL, something which can't be guaranteed in a pure C++ program. The partial flashing service is enabled by default for MakeCode and Python programs.

Reviewer: Do we need to specify that this is the build configuarion of those projects that does this? IE, should we say "Both MakeCode and MicroPython builds currently enable this service by default"

I think it's worth explaining that this service is brought up in the uBit object based on entering pairing mode (assuming I'm right about that). Perhaps the most useful context here is that for both those languages, the service enters the system via the uBit object?

Something like
```
In order to work properly, the partial flashing service requires certain things tobe true about the memory map, and it must be possible for the MemoryMapService to be able to build a model of the flash layout. Most significantly, the service is ideal for updating data outside the main C/C++ build containing CODAL. This is currently possible for MakeCode and MicroPython builds, but not trivial for CODAL programs, where the program is linked into a single executable.

The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is initialised in pairing mode by default, which is implemented as part of the uBit object
```
Perhaps a link to the Pairing Mode check is good? And then is the PartialFlashingService also enabled by default if Bluetooth is enabled? If so what are the config options that are relevant?

Reviewer 2: When configured, the PF service is present in both application and pairing modes.

## Characteristic Commands

> The service consists of a single Bluetooth GATT characteristic that responds to a client's WRITE WITHOUT RESPONSE request with a BLE Notification. The characteristic supports commands to: Read the REGION INFO i.e. request Memory Map information, WRITE DATA to the flash, inform the micro:bit that the sevice has reached END OF TRANSMISSION, obtain the MICROBIT STATUS (PFS version #, current m:b mode), and to RESET the micro:bit into either application (MakeCode) or BLE mode.

Reviewer: Could this be a list please?

# Partial Flashing Service

## Introduction

> The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the firmware. For MakeCode programs this is the user's script, and for Python programs this is the file system.

Reviewer: s/firmware/flash contents/
```suggestion
The Partial Flashing Service allows a BLE client to connect to a micro:bit and read and write the information required to partially update the flash contents. For MakeCode programs this usually means updating the user's script (compiled to asm), and for Python programs this is the entire file system.
```

> This service exists in partnership with the DFU service that allows a client to fully flash the micro:bit.

Reviewer: ```suggestion
This service exists in partnership with the DFU service that allows a client to fully flash the application storage of the micro:bit (for example, for switching between a MicroPython programme and a MakeCode programme).
```

Diagrams showing the flow of data during partial flashing can be found at [the bottom of the document](#partial flashing client).

## Bluetooth Service Specification

The micro:bit v2 follows the same BLE specification as the micro:bit v1. The specification can be found [here](https://lancaster-university.github.io/microbit-docs/resources/bluetooth/bluetooth_profile.html)

## Enabling the service

In order to work properly, the partial flashing service requires certain things to be true about the memory map, and it must be possible for the MemoryMapService to be able to build a model of the flash layout. Most significantly, the service is ideal for updating data outside the main C/C++ build containing CODAL. This is currently possible for MakeCode and MicroPython builds, but not trivial for CODAL programs, where the program is linked into a single executable.

> The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is available in both application and pairing modes.

Reviewer: ```suggestion
The partial flashing service is included by default in MakeCode builds for micro:bit V1 (using microbit-dal, not CODAL) and V2, and MicroPython builds for micro:bit V2 only. The Partial Flashing service is available in both application and pairing modes.

- MakeCode: Enabled V1 (microbit-dal) and V2 (CODAL)
- MicroPython: Enabled on V2 only (CODAL)

```
